---
layout: lesson
root: .  # Is the only page that doesn't follow the pattern /:path/index.html
permalink: index.html  # Is the only page that doesn't follow the pattern /:path/index.html
---


**Welcome** to this lesson on the fundamental principles of **Pangenomics**, a rapidly advancing field in bioinformatics. Throughout this course, you will delve into the basic theories that underpin the study of pangenomes. By utilizing **command-line** software, you will gain hands-on experience in downloading and annotating public bacterial genomes, thereby acquiring essential skills for genomic analysis.  

One of the key highlights of this course is the opportunity to engage with specialized programs designed for pangenomics analysis. You will master the art of **gene family clustering**. You will become adept at constructing **interactive pangenome graphs and plots**, powerful visualization tools for studying the general structure of a pangenome and the families composing it. You will finally explore how to apply **Topological Data Analysis** to the study of pangenomes.

The analyses presented here were meticulously curated to equip you with the necessary tools for conducting a starting pangenomics pipeline. By refining your bioinformatician skills through practical application, you will not only gain confidence in your abilities but also be well-prepared to **explore diverse resources** (See [Other Resources](https://paumayell.github.io/pangenomics/13-other-resources/index.html)). With this, you can go ahead and develop your personalized workflow tailored to the specific objectives of your pangenomics research.   

Get ready to embark on this exciting journey into the world of **Pangenomics**, where you will unlock new insights and unravel the complexities of genomic variation!

<!-- this is an html comment -->

{% comment %} This is a comment in Liquid {% endcomment %}

> ## Pre-requisites
>
> Before diving into this lesson on Pangenomics, it is essential to have a working understanding of the **Bash shell** and the **language Python**. If you are not already familiar with these programming languages, we recommend completing the [Introduction to the Command Line for Pangenomics](https://czirion.github.io/shell-pangenomics/) lesson prior to starting this one and Introduction to Python for Pangenomics FIXME ðŸ’¢. 
> 
> Additionally, some familiarity with biological concepts is assumed for this lesson. It is beneficial to have a basic understanding of **prokaryote**, **genomes**, **genes**, and **orthology**. If you are new to these concepts, we encourage you to review relevant materials to ensure a solid foundation for this course. 
> 
> Throughout this lesson, we will be utilizing data hosted on an **Amazon Machine Instance (AMI)**. Workshop participants will receive information on how to log in to the AMI during the workshop. If you are studying independently, you will need to set up your own AMI or install the necessary programs on your personal computer. Detailed instructions on setting up an AMI and accessing the required data can be found on the [Pangenomics Workshop Setup page](https://czirion.github.io/pangenomics-workshop/setup.html).  
> If you are taking this workshop in **UNAM-CCM** you will access the shell and python, and have access to all the bioinformatics programs through a [JupyterHub server](https://czirion.github.io/pangenomics-workshop/setup.html#connection-to-jupyterhub-notebooks-and-terminal).
>
{: .prereq}

This lesson is the third part of the [Pangenomics Workshop](https://czirion.github.io/pangenomics-workshop/), which also includes [Introduction to the Command Line for Pangenomics](https://czirion.github.io/shell-pangenomics/) and Introduction to Python for Pangenomics FIXME ðŸ’¢.

{% include links.md %}
