---
layout: lesson
root: .  # Is the only page that doesn't follow the pattern /:path/index.html
permalink: index.html  # Is the only page that doesn't follow the pattern /:path/index.html
---

Welcome to this comprehensive lesson on the fundamental principles of Pangenomics, a rapidly advancing field in bioinformatics. Throughout this course, you will delve into the basic theories that underpin the study of Pangenomes. By utilizing command-line software, you will gain hands-on experience in downloading and annotating public bacterial genomes, thereby acquiring essential skills for genomic analysis. 

One of the key highlights on this course is the opportunity to engage with three specialized programs designed specifically for pangenomic analysis. You will master the art of gene family clustering, enabling you to identify commonalities and variations across genomes. Additionally, you will become adept at constructing interactive pangenome graphs, a powerful visualization tool for comprehending the general structure of a pangenome. Furthermore, you will explore the neighborhood of gene families, allowing you to decipher how conserved or variable is the genomic loci where they are encoded. 

The analyses presented here were meticulously curated to equip you with the necessary tools for conducting pangenomic research. By refining your bioinfomatician skills through practical application, you will not only gain confidence in your abilities but also be well-prepared to explore diverse resources (See [other resources](https://paumayell.github.io/pangenomics/07-other-resources/index.html)). Ultimately, you will be able to develop a personalized workflow tailored to your specific pangenomics research objectives. 


Get ready to embark on this exciting journey into the world of Pangenomics, where you will unlock new insights and unravel the complexities of genomic variation!


<!-- this is an html comment -->

{% comment %} This is a comment in Liquid {% endcomment %}

> ## Pre-requisites
>
> Before diving into this lesson on Pangenomics, it is essential to have a working understanding of the bash shell. If you are not already familiar with the bash shell, we recommend completing the [Introduction to the Command Line for Pangenomics](https://czirion.github.io/shell-pangenomics/) lesson prior to starting this one. 
> 
> Additionally, some familiriaty with biological concepts is assumed for this lesson. It is benefitial to have a basic understanding of prokaryote genomes, genes, and orthology. If you are new to these concepts, we encourage you to review relevant materials to ensure a solid foundation for this course. 
> 
> Throughout this lesson, we will be utilizing data hosted on an Amazon Machine Instance (AMI). Workshop participants will receive information on how to log in to the AMI during the workshop. If you are studying independently, you will need to set up your own AMI or install the necessary programs on you personal computer. Detailed instructions on setting up an AMI and accessing the required data can be found on the [Pangenomics Workshop Setup page](https://czirion.github.io/pangenomics-workshop/setup.html).
>
{: .prereq}

{% include links.md %}
