---
layout: lesson
root: .  # Is the only page that doesn't follow the pattern /:path/index.html
permalink: index.html  # Is the only page that doesn't follow the pattern /:path/index.html
---

In this lesson you will learn the basic theory of the study of Pangenomes. You will download and annotate public genomes using command-line software. 
Most importantly you will conduct pangenomic analysis with three specialized programs, which include clustering of gene families, construction of 
interactive pangenome graphs and visualize neighbournig gene families.

The analyses presented here are intended to give you the necessary tools to carry out pangenomic analyses, as well as allow you to practice your 
bioinformatician skills. This experience will let you to explore other resources and end up with your personalized workflow for your pangenomics research. 

<!-- this is an html comment -->

{% comment %} This is a comment in Liquid {% endcomment %}

> ## Prerequisites
>
> This lesson assumes a working understanding of the bash shell. If you haven’t already 
> completed the lesson [Introduction to the Command Line for Pangenomics](https://czirion.github.io/shell-pangenomics/), and you aren’t 
> familiar with the bash shell; please review those materials before starting this lesson.
>
> This lesson also assumes some familiarity with biological concepts, 
> such as prokaryote genome, gene and orthology.
> 
> This lesson uses data hosted on an Amazon Machine Instance (AMI). Workshop participants
> will be given information on how to log in to the AMI during the workshop. Learners using 
> these materials for the self-directed study will need to set up their own AMI or install programs in their personal computers. Information 
> on setting up an AMI and accessing the required data is provided on the 
> [Pangenomics Workshop Setup page](https://czirion.github.io/pangenomics-workshop/setup.html).
>
{: .prereq}

{% include links.md %}
